File: api/routers/players.py
"""
Players Router
===============
Defines all HTTP endpoints related to player data, predictions, and metadata.

FastAPI uses Python type hints to:
  1. Parse and validate incoming data automatically
  2. Generate /docs documentation without any extra work
  3. Return helpful 422 errors when data is wrong

ENDPOINT OVERVIEW:
  GET /api/v1/players/games            → list of supported game platforms (from game_registry)
  GET /api/v1/players/models           → list of registered ML models (from model_registry)
  GET /api/v1/players                  → browse/search players in the dataset
  GET /api/v1/players/{platform}/{id}  → full analytics for one player

NOTE ON ROUTE ORDER:
FastAPI matches routes top to bottom. "/games" and "/models" must be defined
BEFORE "/{platform}/{id}" otherwise FastAPI would try to interpret "games"
as a platform parameter.

LEARN MORE:
  Path parameters:  https://fastapi.tiangolo.com/tutorial/path-params/
  Query parameters: https://fastapi.tiangolo.com/tutorial/query-params/
  HTTPException:    https://fastapi.tiangolo.com/tutorial/handling-errors/
"""

from fastapi import APIRouter, HTTPException, Query

from api.registry.game_registry import get_supported_games, get_game
from api.registry.model_registry import list_models, DEFAULT_MODEL
from api.services.data_service import get_player, list_players
from api.services.model_service import predict_churn
from api.services.shap_service import get_player_shap

# APIRouter groups related endpoints.
# Registered in main.py with prefix="/api/v1/players".
router = APIRouter()


@router.get("/games")
def get_games():
    """
    Returns the list of supported gaming platforms.
    The React frontend calls this to dynamically build the platform dropdown.
    When you add a new game to game_registry.py, it appears here automatically.
    """
    return get_supported_games()


@router.get("/models")
def get_models():
    """
    Returns all registered ML models with descriptions.
    The frontend can use this to let users choose which model powers predictions.
    """
    return list_models()


@router.get("")
def search_players(
    platform: str | None = Query(default=None, description="Filter by platform ID (e.g., chess_com)"),
    limit: int = Query(default=50, ge=1, le=500, description="Max number of players to return"),
):
    """
    Browse players in the dataset, optionally filtered by platform.

    Query parameters are parsed automatically from the URL:
      GET /api/v1/players?platform=chess_com&limit=20

    TODO: Implement this endpoint.
    Steps:
      1. Call list_players(platform=platform, limit=limit) from data_service
      2. Wrap in a try/except — if the features file doesn't exist, raise:
         HTTPException(status_code=503, detail="Features not ready. Run `make train`.")
      3. Return the list of player dicts
    """
    raise NotImplementedError("TODO: implement search_players()")


@router.get("/{platform}/{player_id}")
def get_player_analytics(
    platform: str,
    player_id: str,
    model_id: str = Query(default=DEFAULT_MODEL, description="Model ID to use for prediction"),
):
    """
    Core endpoint: returns full analytics for a single player.

    This powers the Player Lookup page in the React frontend.

    Response shape:
    {
        "player_id": "hikaru",
        "platform": "chess_com",
        "features": { ...all feature columns... },
        "prediction": {
            "churn_probability": 0.73,
            "churn_predicted": true,
            "risk_level": "High",
            "model_used": "ensemble"
        },
        "shap_values": [
            {"feature": "days_since_last_game", "label": "...", "shap_value": 0.42, "direction": "increases_churn"},
            ...
        ]
    }

    TODO: Implement this endpoint.
    Steps:
      1. Validate platform: try get_game(platform), raise HTTPException(404) if KeyError
      2. Look up player: features = get_player(player_id, platform)
         If None: raise HTTPException(status_code=404, detail="Player not found in dataset.")
      3. Run prediction: prediction = predict_churn(features, model_id)
         Catch any exception and raise HTTPException(500) with the error message
      4. Get SHAP values: shap = get_player_shap(player_id, platform)
         (shap can be None — that's ok, just include it as-is)
      5. Return the combined dict in the shape above
    """
    raise NotImplementedError("TODO: implement get_player_analytics()")

File: frontend/src/api/players.js
/**
 * Players API Module
 * ===================
 * All API calls related to players, games, and models live here.
 *
 * WHY SEPARATE API FUNCTIONS FROM COMPONENTS?
 * --------------------------------------------
 * Components should only care about rendering — not URLs, HTTP methods,
 * or response shapes. These functions are the contract between your UI
 * and your backend.
 *
 * If the backend changes an endpoint (e.g., adds a /v2 prefix), you update
 * this file only — no component needs to change.
 *
 * HOW TO USE THESE IN A COMPONENT:
 * ----------------------------------
 * Direct (in a useEffect):
 *   const data = await fetchPlayerAnalytics('chess_com', 'hikaru')
 *
 * Better — use the usePlayer() custom hook, which wraps these functions
 * and automatically manages loading/error state for you.
 * See src/hooks/usePlayer.js.
 */

import client from './client'

/**
 * Fetch the list of supported gaming platforms.
 * Powers the platform dropdown in PlayerSearch — fetched dynamically
 * so adding a new game to the backend registry appears without a frontend deploy.
 *
 * Returns: [{ id: "chess_com", display_name: "Chess.com", player_id_example: "hikaru", ... }]
 *
 * TODO: Implement using client.get('/api/v1/players/games')
 * Return response.data
 */
export async function fetchSupportedGames() {
  throw new Error('TODO: implement fetchSupportedGames()')
}

/**
 * Fetch all registered ML models.
 * Use this to populate a model selector (if you want users to pick a model).
 *
 * Returns: [{ id: "ensemble", display_name: "Soft-Voting Ensemble", description: "..." }]
 *
 * TODO: Implement using client.get('/api/v1/players/models')
 */
export async function fetchModels() {
  throw new Error('TODO: implement fetchModels()')
}

/**
 * Fetch full analytics for one player. This is the main call.
 * Powers the entire AnalyticsPanel and provides context to the ChatPanel.
 *
 * @param {string} platform  - e.g., "chess_com"
 * @param {string} playerId  - e.g., "hikaru"
 * @param {string} modelId   - e.g., "ensemble" (optional)
 *
 * Returns:
 * {
 *   player_id:   "hikaru",
 *   platform:    "chess_com",
 *   features:    { games_7d: 14, engagement_score: 72.3, ... },
 *   prediction:  { churn_probability: 0.18, risk_level: "Low", ... },
 *   shap_values: [{ feature: "days_since_last_game", shap_value: -0.31, direction: "decreases_churn" }, ...]
 * }
 *
 * TODO: Implement using client.get(`/api/v1/players/${platform}/${playerId}`)
 * Pass modelId as a query param: { params: { model_id: modelId } }
 * Return response.data
 */
export async function fetchPlayerAnalytics(platform, playerId, modelId = 'ensemble') {
  throw new Error('TODO: implement fetchPlayerAnalytics()')
}

/**
 * Browse players in the dataset. Optionally filter by platform.
 *
 * TODO: Implement using client.get('/api/v1/players', { params: { platform, limit } })
 */
export async function fetchPlayers(platform = null, limit = 50) {
  throw new Error('TODO: implement fetchPlayers()')
}
